raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -370,8 +363,52 @@ private void maybeFireLeaderChange() {
         }
     }
 
-    @Override
-    public void initialize() {
+    public void initialize(
+        Map<Integer, InetSocketAddress> voterAddresses,
+        String listenerName,
+        QuorumStateStore quorumStateStore,
+        Metrics metrics
```

Comment:
Why do we need to delay initialization of metrics?

I can feel that the parameters for this method might grow over time. Perhaps we should package it up into some kind of initialization object.

Reply:
Yeah. I am also not happy with this move. We only do this delay initialization because of integration tests (`QuorumTestHarness`, `KRaftClusterTestKit`). This is not needed by `**/src/main`.

Once we have KIP-853 fully implemented, I should be able to fix the integration tests to not use the static voter set and the delayed initialization.

I create [Remove delayed initialization because of static voter set](https://issues.apache.org/jira/browse/KAFKA-16653) to track this work.